### PR TITLE
WIP: project updates

### PR DESCRIPTION
**Pull request by sigrlami**
This PR provides a basic outline for different improvements we have for Slick. Includes following corrections:
- Addition of `Serve.hs` that include functionality for `preview` mode that spins out small server for site observation
- Additional integration with our `shake-watch` project that will be released soon which provides continuous rebuilds on either `haskell` changes or `site` changes.
- Addition of `Build.hs` file that includes high-level goodies we created while working on multiple sites
- Updates example project
    - `example-site` moved to `example` directory to accommodate multiple approaches to build site. We can have `minimalist version` and `advanced version` with support of automatic preview in browser and file watch. 
    - Added additional keys related to `preview` and `watch` functionality
    - Additional `Builder.hs` file with everything related to builds and update to the main workflow where shakeRules and builder moved to separate functions for further reuse.
- Compacts `README.md` to concentrate on go-to production details (should be improved in later updates) and updates dependencies.

### CODE REVIEW

File: src/Slick/Build.hs [hunk 1]
```diff
@@ -65,3 +70,13 @@ flattenMeta (Meta meta) = toJSON $ fmap go meta
...
+
+--------------------------------------------------------------------------------
+
+-- | Custom cleanup for the Shake builder
+--
+pruner :: [FilePath] -> IO ()
+pruner listOfFiles = do
+  present <- listFilesRecursive "public"
+  mapM_ removeFile $ (toStandard <$> present) L.\\ (toStandard <$> listOfFiles)
+  putStrLn "PRUNER  "
```
Comment (mgajda) [inline src/Slick/Build.hs]: Can you just inform the user about pruned files here?
```haskell
putStrLn $ "Pruned stale outputs: " <> removedFiles
```

File: src/Slick/Build.hs [hunk 2]
```diff
@@ -65,3 +70,13 @@ flattenMeta (Meta meta) = toJSON $ fmap go meta
   go (MetaString  m) = toJSON m
   go (MetaInlines m) = toJSON $ stringify m
   go (MetaBlocks  m) = toJSON $ stringify m
+
+--------------------------------------------------------------------------------
+
+-- | Custom cleanup for the Shake builder
```
Comment (mgajda) [inline src/Slick/Build.hs]: Better documentation:
```
-- | Remove all files that are in the target directory, but not in the `shake` rules.
```

File: src/Slick/Utils.hs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import           Development.Shake.FilePath
+import           System.Directory
+import           System.Directory.Extra     (listFilesRecursive, removeFile)
+import           System.Environment
+
+------------------------------------------------------------------------------
+-- Helper functions
+
+-- | Generalized sorting function
+sortByM :: (Monad m, Ord k) => (a -> m k) -> [a] -> m [a]
```
Comment (mgajda) [inline src/Slick/Utils.hs]: Is it still used? Looks like weed to me.

File: package.yaml [hunk 4]
```diff
@@ -28,6 +28,8 @@ dependencies:
 - warp
```
Comment (mgajda) [inline package.yaml:40]: Add comment indicating which of these are needed for `--preview`/`--watch`:
```haskell
# Only needed when live building is used:
```

File: README.md [hunk 5]
```diff
@@ -25,15 +25,15 @@ Here's a quick overview of what Slick can do:
     -   Write your blog posts in Markdown or LaTeX and render it to
         syntax-highlighted HTML!
     -   Slick processes Pandoc (and LaTeX) metadata into a usable form (as an
-        Aeson Value object) which you can manipulate as you please.
+        [Aeson](https://hackage.haskell.org/package/aeson) Value object) which you can manipulate as you please.
```
Comment (mgajda) [inline README.md:28]: We used `yaml` to parse metadata, so please link to either `yaml` package and `aeson` `Value` datatype explicitly.

File: example/app/Main.hs [hunk 6]
```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE CPP                        #-}
```
Comment (mgajda) [inline example/app/Main.hs]: Do we still use `CPP`? If not, then `weeder` again :-)

File: src/Slick.hs [hunk 7]
```diff
@@ -1,13 +1,8 @@
...
-    compileTemplate'
+-- | Slick
+--  This module re-exports everything you need to use Slick
 
-  -- ** Pandoc
-  , PandocReader
+module Slick
+  ( PandocReader
   , PandocWriter
   , markdownToHTML
```
Comment (mgajda) [inline src/Slick.hs]: It might be good idea to add my code that allows to use arbitrary options for Pandoc to default example?

File: src/Slick/Caching.hs [hunk 8]
```diff
@@ -63,10 +64,10 @@ jsonCache' loader = unpackJSON
 
 -- | A wrapper around 'jsonCache' which simplifies caching of values which do NOT
 -- depend on an input parameter. Unfortunately Shake still requires that the
--- key type implement several typeclasses, however this is easily accomplished 
+-- key type implement several typeclasses, however this is easily accomplished
```
Comment (mgajda) [inline src/Slick/Caching.hs]: It looks to me like your editor adds a lot of irrelevant spaces in random places.
Could you manage to somehow remove these changes from diff?

File: src/Slick/Pandoc.hs [hunk 9]
```diff
@@ -1,47 +1,64 @@
...
- where
-  exts = mconcat
-    [ extensionsFromList
-      [ Ext_yaml_metadata_block
-      , Ext_fenced_code_attributes
-      , Ext_auto_identifiers
-      ]
-    , githubMarkdownExtensions
-    ]
+markdownOptions =
```
Comment (mgajda) [inline src/Slick/Pandoc.hs]: I believe this should be configurable by default. (And default `markdownOptions` should be kicked to `example/`.)

File: src/Slick/Pandoc.hs [hunk 10]
```diff
@@ -50,76 +67,79 @@ unPandocM p = do
...
+  loadUsing reader writer text >>= convert
+
+-- | Generalized Entity Loader
+--
+loadEntity :: FromJSON b => EntityFilePath a
+           -> Action b
+loadEntity (EntityFilePath path) =
+  subloadEntity path
+
+subloadEntity :: FromJSON b => FilePath -> Action b
```
Comment (mgajda) [inline src/Slick/Pandoc.hs]: Can you comment on what `subloadEntity` does?

File: src/Slick/Serve.hs [hunk 11]
```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Slick.Serve ( runServerStatic
```
Comment (mgajda) [inline src/Slick/Serve.hs]: Missing:
1. A single interface function to start both servers.
2. A single interface function to kill the servers when the process exits.

Since purpose of this module is to provide self-contained testing functionality, we do not need to expose WS and HTTP servers separately.

File: stack.yaml [hunk 12]
```diff
@@ -1,65 +1,2 @@
...
-# arch: i386
-# arch: x86_64
-#
-# Extra directories used by stack for building
-# extra-include-dirs: [/path/to/dir]
-# extra-lib-dirs: [/path/to/dir]
-#
-# Allow a newer minor version of GHC than the snapshot specifies
-# compiler-check: newer-minor
+ghc-options: {"$locals": -ddump-to-file -ddump-hi}
```
Comment (mgajda) [inline stack.yaml]: Do we still need these dumps (looks like unnecessary change to me.)

**Review by mgajda — CHANGES_REQUESTED** [review]
Looks good overall.
Beside issues indicated below, can you please also fix the following:
* revert unnecessary changes of whitespace,
* and unnecessary renaming of `example-site` to `example`. This is not our code, so better be nice to current maintainer by limiting ourselves to functional changes. (Otherwise this diff is unreadable. Out of 119-single-file changes, it seems mostly renamings and whitespace!)

File: stack.yaml (hunk 12 above)
Comment (sigrlami) [reply stack.yaml]: @mgajda unfortunately yes, in version 2 of stack there is regression and I can clearly see it, look issue https://github.com/ndmitchell/weeder/issues/53

File: src/Slick/Caching.hs (hunk 8 above)
Comment (sigrlami) [reply src/Slick/Caching.hs]: Yes, I'll try to revert them. I use regular Emacs. It's actually deletes spaces aggressively, not sure what's wrong with comments.

File: README.md (hunk 5 above)
Comment (sigrlami) [reply README.md:28]: @mgajda this is not my change intoduced by other commiter to upstream repo https://github.com/ChrisPenner/slick/commit/f7a6c5205319a8359b610ccf6330c65737b834d0

File: src/Slick/Build.hs (hunk 2 above)
Comment (sigrlami) [reply src/Slick/Build.hs]: Fixed

File: src/Slick/Utils.hs (hunk 3 above)
Comment (sigrlami) [reply src/Slick/Utils.hs]: @mgajda removed as weed

File: src/Slick/Pandoc.hs (hunk 10 above)
Comment (sigrlami) [reply src/Slick/Pandoc.hs]: @mgajda description added

File: src/Slick/Build.hs (hunk 1 above)
Comment (sigrlami) [reply src/Slick/Build.hs]: Added informational message about removed files

File: src/Slick/Pandoc.hs (hunk 10 above)
Comment (sigrlami) [reply src/Slick/Pandoc.hs]: comment added

**Comment by sigrlami** [comment]
@mgajda unnecessary renaming of example-site to example removed, now only 19 files in patchset

File: src/Slick.hs (hunk 7 above)
Comment (sigrlami) [reply src/Slick.hs]: Not sure which code you're referring here. Can you elaborate more?

File: src/Slick/Serve.hs (hunk 11 above)
Comment (sigrlami) [reply src/Slick/Serve.hs]: So, here is a tricky thing, `runServerWS` not only runs websocket server but also an http server too. So we can consider it as (1). I wanted to have also separate options to run just http server for some users, but if you think that superfluous I can remove `runServer` and `runServerStatic`

Comment (mgajda) [reply src/Slick/Serve.hs]: Oh, then please rename them as well.
It looks confusing at the moment.

Just a single `runServer` with two port parameters (for websockets and HTTP) is sufficient.

File: README.md (hunk 5 above)
Comment (mgajda) [reply README.md:28]: Would be good to revert whitespace changes to avoid confusion.

File: package.yaml (hunk 4 above)
Comment (sigrlami) [reply package.yaml:40]: Rearranged, grouped everything for preview into single block, added comment.

File: src/Slick/Serve.hs (hunk 11 above)
Comment (sigrlami) [reply src/Slick/Serve.hs]: Interface changed to 3 functions `serverStart`, `serverStop`, `serverHandler` last one needed for handling request coming to server. Due to nature of Wai there is no simple way to stop server, one need to track MVar and depend on it in order to kill server. Not sure if full example required in `example-site` as it creates additional layer.

File: README.md (hunk 5 above)
Comment (sigrlami) [reply README.md:28]: I checked again, my PR actually strips down dangling spaces. And it's default for many editors, initial commits actually have spaces. I don't think that will create any confusion, there are just 3 places where it's visible.

File: src/Slick/Pandoc.hs (hunk 9 above)
Comment (sigrlami) [reply src/Slick/Pandoc.hs]: so, better to remove completely?

**Comment by sigrlami** [comment]
Restored original whitespaces and alignments, now PR stripped down to 14 files.

File: example/app/Main.hs (hunk 6 above)
Comment (sigrlami) [reply example/app/Main.hs]: CPP removed

File: src/Slick/Utils.hs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  concat <$> mapM getPaths dirs
+    where
+      getPaths :: FilePath -> Action [FilePath]
+      getPaths dir =
+        fmap (dir </>) <$>
+          getDirectoryFiles dir extensions
+
+-- | Sequential composition of monadic functions
+mseq :: Monad m => (a -> m b) -> (b -> m c) -> a -> m c
+f `mseq` g =  \x -> f x >>= g
```
Comment (mgajda) [inline src/Slick/Utils.hs]:
```
import Control.Monad
(>=>) :: Monad m => (a -> m b) -> (b -> m c) -> a -> m c
```

File: example-site/app/Main.hs [hunk 14]
```diff
@@ -1,130 +1,111 @@
...
+  [ Option "" ["preview"      ] (NoArg $ Right Preview    ) "running as preview"
+  , Option "" ["watch"        ] (NoArg $ Right Watch      ) "running with watch"
+  ]
+
+-- | Specific build rules for the Shake system
+--   defines workflow to build the webiste
+buildRules :: Foldable t => t Flags -> Rules ()
+buildRules flags = do
+
+  let isPreviewMode = Preview `elem` flags
```
Comment (mgajda) [inline example-site/app/Main.hs:60]: Thought we agreed that `--preview` and `--watch` should be a single CLI option that starts different things inside?

File: example-site/app/Main.hs [hunk 15]
```diff
@@ -1,130 +1,111 @@
...
+  "dist/posts//*.html" %> buildPost postCache
+
+
+-- | Function to start custom Shake pipeline for execution
+--
+runSiteBuilder :: ShakeOptions                     -- ^ Options for the Shake builder
+               -> [OptDescr (Either String Flags)] -- ^ Converted CLI arguments
+               -> IO ()
+runSiteBuilder shOpts flags =
+  shakeArgsPruneWith shOpts pruner flags $
```
Comment (mgajda) [inline example-site/app/Main.hs]: I believe you still need to add `--prune` here so that Shake will *always* prune.
(This is not necessary in other applications, but important for usability here.)

**Review by mgajda — CHANGES_REQUESTED** [review]
Three remaining issues, others successfully resolved.

Looks okay to me, but I still request you to fix these.

File: src/Slick/Utils.hs (hunk 13 above)
Comment (sigrlami) [reply src/Slick/Utils.hs]: >=> restored

File: example-site/app/Main.hs (hunk 14 above)
Comment (sigrlami) [reply example-site/app/Main.hs:60]: `watch` removed; `prune` added instead

File: example-site/app/Main.hs (hunk 15 above)
Comment (sigrlami) [reply example-site/app/Main.hs]: added `--prune` and resolver to pick correct `shakeBuildWith`

**Review by sigrlami — COMMENTED** [review]
`--prune`

File: example-site/app/Main.hs (hunk 14 above)
Comment (mgajda) [reply example-site/app/Main.hs:60]: I sent you a message suggesting other way of doing that.

File: src/Slick/Pandoc.hs [hunk 16]
```diff
@@ -59,8 +59,11 @@ markdownToHTML' :: (FromJSON a)
                 -> WriterOptions  -- ^ Pandoc writer options to modify output
                 -> T.Text         -- ^ Text for conversion
                 -> Action a
-markdownToHTML' rops wops txt =
-  mseq (markdownToHTML rops wops) convert txt
+markdownToHTML' rops wops =
+  (markdownToHTML rops wops) >=> convert
```
Comment (mgajda) [inline src/Slick/Pandoc.hs]: You may remove superfluous parenthesis, since function application has always higher priority

Comment (sigrlami) [reply src/Slick/Pandoc.hs]: removed

File: example-site/app/Main.hs (hunk 14 above)
Comment (sigrlami) [reply example-site/app/Main.hs:60]: Unfortunately, adding `--prune` by hand won't work, looking at the code Shake stores `stdin` reference on their own and adds `OptsDescr Prune` .. to flags and tries to read flags [shakeArgsPrune](https://hackage.haskell.org/package/shake-0.18.3/docs/src/Development.Shake.Util.html#shakeArgsPrune).

**Review by mgajda — APPROVED** [review]
Looks good for our use case.
Now it is up to Chris to look at it for inclusion upstream.
